k8s/src/main/scala/io/buoyant/k8s/IngressCache.scala

```diff
@@ -0,0 +1,188 @@
+    }
+  }
+
+  def add(obj: v1beta1.Ingress): Unit =
+    for (item <- mkItem(obj)) synchronized {
+      val name = getName(obj).get //todo
+      log.debug("k8s ns %s added: %s", namespace.getOrElse(""), name)
+      val items = state.sample() match {
+        case Activity.Ok(items) => items
+        case _ => Map.empty[NamespacedName, VarUp[IngressSpec]]
+      }
+      state() = Activity.Ok(items + (name -> Var(item)))
+    }
+
+  def modify(obj: v1beta1.Ingress): Unit =
+    for (name <- getName(obj)) synchronized {
+      log.debug("k8s ns %s modified: %s", namespace.getOrElse(""), name)
+      state.sample() match {
+        case Activity.Ok(snap) =>
+          snap.get(name) match {
+            case None =>
+              log.warning("k8s ns %s received modified watch for unknown resource %s", namespace.getOrElse(""), name)
+            case Some(item) =>
+              mkItem(obj) match {
+                case Some(i) => item() = i
+                case None => state() = Activity.Ok(snap - name)
+              }
+          }
+        case _ =>
+      }
+    }
+
+  def delete(obj: v1beta1.Ingress): Unit =
+    for (name <- getName(obj)) synchronized {
+      log.debug("k8s ns %s deleted: %s", namespace.getOrElse(""), name)
+      state.sample() match {
+        case Activity.Ok(snap) =>
+          for (svc <- snap.get(name)) {
+            state() = Activity.Ok(snap - name)
+          }
+
+        case _ =>
+      }
+    }
+
+  def update(watch: v1beta1.IngressWatch): Unit = watch match {
+    case v1beta1.IngressError(e) => log.error("k8s watch error: %s", e)
+    case v1beta1.IngressAdded(ingress) => add(ingress)
+    case v1beta1.IngressModified(ingress) => modify(ingress)
+    case v1beta1.IngressDeleted(ingress) => delete(ingress)
+  }
+
+  /**
+   * Ingress-specific methods
+   */
+
+  def getName(ingress: v1beta1.Ingress): Option[NamespacedName] =
+    ingress.metadata.flatMap { i =>
+      (i.namespace, i.name) match {
+        case (Some(ns), Some(n)) => Some(NamespacedName(i.namespace, i.name))
+        case _ => None
+      }
+    }
+
+  def mkItem(ingress: v1beta1.Ingress): Option[IngressSpec] = {
+    val namespace = ingress.metadata.flatMap(meta => meta.namespace).getOrElse("default")
+    ingress.spec match {
+      case Some(spec) =>
+        val paths = for (
+          spec <- ingress.spec.toSeq;
+          rules <- spec.rules.toSeq;
+          rule <- rules;
+          http <- rule.http.toSeq;
+          path <- http.paths
+        ) yield {
+          IngressPath(rule.host, path.path.map(Path.read(_)), namespace, path.backend.serviceName, path.backend.servicePort)
+        }
+
+        val fallback = spec.backend.map(b => IngressPath(None, None, namespace, b.serviceName, b.servicePort))
+        Some(IngressSpec(ingress.metadata.flatMap(meta => meta.name), fallback, paths))
+      case None => None
+    }
+  }
+
+  def isPrefix(pfx: Path, str: Path): Boolean = pfx.showElems == str.take(pfx.size).showElems
```

Comment:
Thanks for looking at this. This was my attempt to treat prefixes the same way that the path identifier does. There are two controllers implemented by the kubernetes team (GCE & nginx) and they themselves disagree on the behaviour.

See https://github.com/kubernetes/contrib/issues/885